File: C++/0295-find-median-from-data-stream.cpp
// Tags: Amazon Heap Design
// Time:  O(logn) per addNum, O(1) per findMedian.
// Space: O(n), total space
// Two Heaps solution
class MedianFinder {
private:
    priority_queue<int, vector<int>, less<int>> max_heap;
    priority_queue<int, vector<int>, greater<int>> min_heap;

public:
    /** initialize your data structure here. */
    MedianFinder() {

    }

    void addNum(int num) {
        // only check num > top if max_heap is not empty
        // or if (min_heap.empty() || num > min_heap.top())
        if (max_heap.empty() || num > max_heap.top()) {
            min_heap.emplace(num);
            if (min_heap.size() > max_heap.size() + 1) {
                max_heap.emplace(min_heap.top());
                min_heap.pop();
            }
        } else {
            max_heap.emplace(num);
            if (max_heap.size() > min_heap.size()) {
                min_heap.emplace(max_heap.top());
                max_heap.pop();
            }
        }
    }

    double findMedian() {
        return min_heap.size() == max_heap.size() ?
            (max_heap.top() + min_heap.top()) / 2.0 : min_heap.top();
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */


// Notes:
// max_heap covers first half and min_heap covers the second half.
// Our balancing strategy is to keep min_heap with size either equal
// or 1 greater than max_heap size.


// Time: O(logn + n) ~ O(n), per addNum
// Binary search takes logn, whereas insertion can take upto n as
// elements have to be shifted.
// Space: O(n)
// Insertion sort solution
class MedianFinder {
private:
    vector<int> data;
public:
    /** initialize your data structure here. */
    MedianFinder() {

    }

    void addNum(int num) {
        if (data.empty())
            data.push_back(num);
        else
            data.insert(data.begin() + search(num)+1, num);
            // or data.insert(lower_bound(data.begin(), data.end(), num), num);
    }

    double findMedian() {
        int n = data.size();
        return n & 1 ? data[n/2] : ((double) data[n/2 - 1] + data[n/2])/2;

    }
    int search(int num) {
        int left = 0, right = data.size() - 1, ans = -1;
        while (left <= right) {
            int mid = left + (right - left)/2;
            if (data[mid] <= num) {
                ans = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return ans;
    }
};

File: Python/0295-find-median-from-data-stream.py
# Time:  O(nlogn) for total n addNums, O(logn) per addNum, O(1) per findMedian.
# Space: O(n), total space
# Heap Solution
import heapq
# or from heapq import heappush, heappop
class MedianFinder:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min_heap = []
        self.max_heap = []


    def addNum(self, num: int) -> None:
        if not self.max_heap or num > -self.max_heap[0]:
            heapq.heappush(self.min_heap, num)
            if len(self.min_heap) > len(self.max_heap) + 1:
                heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))
        else:
            heapq.heappush(self.max_heap, -num)
            if len(self.max_heap) > len(self.min_heap):
                heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))

    def findMedian(self) -> float:
        return (self.min_heap[0] - self.max_heap[0])/2 if len(self.min_heap) == len(self.max_heap)\
                else self.min_heap[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
